### ISSUE TITLE
[Perf][KV Group] 两份分组方案缺少 GDN 实测耗时：共享缓存零代价且优于 group_size=16

### ISSUE THREAD

**Issue by ChunnX** [singzhou/vllm-ascend#4]
对 `70f1d583`（codex plan）与 `0c187fce`（dsv4 review）两份方案的综合评审。

两份文档的工程严谨度都很高，尤其是 dsv4 review 的逐条源码核实（`use_dspark()` 存在性、短组下游容忍、容量推导、core 是 scheduler/worker 共用入口）不是空口背书，这些核实结论我复核后全部成立。下面的意见不否定这些工作，而是指出**一个两份文档都没有掌握的实测数据，它改变了优先级判断**。

## 1. 两份计划都缺少 GDN 逐次耗时数据

在 `kv_spec_recover/plan/` 两份文档里 grep `GDN` / `14ms` 均无命中。codex plan §2.2 很克制地承认了这一点：

> DSpark 日志没有记录 `[DIAG-BUILD-META-V2]` 和逐次耗时，因此不能把"3.75 倍构建单元"直接写成"实测耗时精确增加 3.75 倍"。

这个自我约束是对的。现在这个数据有了（profiling 实测）：

```
_build_attention_metadata 总耗时  ≈ 17ms
  其中 GDN builder build         ≈ 14ms   (10 次调用，每次 ~1.4ms)
  其余（含 15 次 metadata 构造）  ≈  3ms
```

**GDN 占 82%。** 而单次 1.4ms 里绝大部分是重复计算。

`vllm_ascend/ops/gdn_attn_builder.py` 的 `build()` 全长约 358 行（408-766），从 common metadata **总共只读 5 个字段**：

| 读取项 | 随 group 变? |
|---|---|
| `query_start_loc` / `query_start_loc_cpu` | 否 |
| `compute_num_computed_tokens()` | 否 |
| `seq_lens` | 否 |
| `block_table_tensor` | **是** |

`block_table_tensor` 只出现在 6 处，除第一处 `mamba_get_block_table_tensor()` 外全是切片：

```python
block_table_tensor = mamba_get_block_table_tensor(...)      # 唯一实际计算
non_spec_state_indices_tensor = block_table_tensor[:, 0]     # 切片
non_spec_conv1d_cache_indices = block_table_tensor           # 引用
... block_table_tensor[:, : self.num_spec + 1]               # 切片 x2
... block_table_tensor[:, 0]                                 # 切片
```

昂贵的部分（`split_decodes_and_prefills`、spec mask、多次 `torch.zeros/empty/arange` 设备分配、`cumsum`、`index_select`）**全部与 group 无关**。

结论：改分组只能减少调用**次数**，动不了单次调用内部的冗余。即使降到 5 组，仍有 4 次在重算同一批次形状。

## 2. 综合判断：共享缓存方案在每个维度上都占优

所谓共享缓存，是在 `build_attn_metadata` 的 group 循环**外**建一个 batch 作用域缓存、逐组传给 GDN builder，命中则跳过全部 group-无关计算，只重算 `mamba_get_block_table_tensor` 与那 5 个切片。

**这不是"合并 metadata"。** 两份文档的 §4.4 都以"不同 group 有独立 block table / slot mapping，私自合并与 scheduler 分配不一致"否决了 worker 侧合并——理由正确，但不适用于共享缓存：每个 group 照样产出完整独立的 metadata，block table 与 slot mapping 一个不动，与 scheduler 的映射关系完全不变，被复用的只是 builder 内部的中间批次形状计算。

而且 `vllm_ascend/worker/v2/attn_utils.py` 的**同一个函数里已有同构先例**——DSA 的三个共享 dict：

```python
# DSA metadata is shared by the ratio-specific cache groups for one model
# execution. Keep the cache at the batch-builder scope,
# so each DSA builder can reuse the split results and SAS metadata produced
# by earlier groups in this invocation.
prefill_ratio_to_sas_metadata: dict[Any, Any] = {}
decode_ratio_to_sas_metadata: dict[Any, Any] = {}
common_ratio_to_sas_metadata: dict[Any, Any] = {}
```

GDN 需要的是同一个机制。对比：

| 维度 | group_size=16 方案 | 共享缓存 |
|---|---|---|
| GDN 耗时 | 3 次 × 1.4 ≈ 4.2ms | 1 次完整 + 9 次切片 ≈ **1.5~2ms** |
| 总耗时 | ~7ms | **~5ms** |
| KV 容量 | **-6.25%**（186,413 → ~174,762） | **0** |
| 改动落点 | **vLLM core**（dsv4 §5 定为"第一闸口"：须上游 PR 或 fork patch） | 仅 vllm-ascend 两个文件 |
| 头号失败风险 | scheduler/worker 分组不一致 → block table 越界 | **无**（组数不变，两侧不可能分歧） |
| 上游依赖 | 有 | 无 |

**dsv4 review 自己指出的两点，恰好放大了这个差距**：把 core 改动提升为"第一闸口"（上游 PR 或长期维护 fork patch），以及把 scheduler/worker 一致性列为"头号失败风险"。共享缓存两者都不涉及。

## 3. 头号风险其实来自 gating，不是来自改 group_size

dsv4 §0 保留意见 2：

> gating 依赖 `vllm_config.speculative_config.use_dspark()` 在两处都返回 True。若任一侧 `speculative_config` 为空，会出现 15 组 vs 5 组不一致，直接导致 block table / slot_mapping 越界。

这个风险**完全来自 `use_dspark()` gating 本身**，而不是 group_size 的取值。若改成不带 gating 的通用规则，scheduler 与 worker 从同一份 spec 出发必然算出同一结果，该风险由构造消除。

两份文档都以"影响所有 hybrid 模型、不可控"为由否决全局改动——这对 `MIN_GROUP_SIZE=16` 这类固定常数成立，但对下面这条规则不成立：

> **组数最少，且 padding 不超过现有 heuristic 的 padding。**

现状永远是候选之一，所以该规则**在任何模型上都不会比今天差**，是全局安全的严格改进；同时它正是 vLLM 自己 FIXME 承认待解决的问题：

> `_get_kv_cache_groups_uniform_page_size`:
> Need a better strategy if we want to support more complex patterns
> (e.g., 20 full + 30 sw, where the group size should be 10).

做成通用规则既消除头号风险，又更容易上游化。

## 4. 若要改分组，目标应是 g=8 而非 g=16

两份文档都把 `g=8` 放在回退位置（dsv4 §10："容量损失超 10% 时用 group_size=8，理论上不损容量"），但按 codex §4.2 自己的表格：

| group_size | 组数 | padding 比例 | 相对现状 |
|---|---|---|---|
| 5（现状） | 15 | 8.70% | — |
| **8** | **9** | **4.35%** | 组数↓、容量↑（75/72），**纯赚** |
| 16（推荐） | 5 | 15.94% | 组数↓、容量↓ 6.25% |

`g=8` 在两个维度上都优于现状。选 `g=16` 的隐含前提是"组数越少越值得花容量买"——**有了共享缓存后该前提不成立**：组数 9→5 只再省约 0.4ms，却要付 6.25% KV 容量 + core 改动 + gating 风险。

另外 codex plan 内部有一处不自洽：§4.1 的搜索上限 `DSPARK_KV_GROUP_MAX_PADDING_RATIO = 0.20` 比 §10.2 的验收门槛"容量损失不超过 10%"宽一倍，意味着搜索可能选出必然验收失败的候选。若改用"padding ≤ 现有 heuristic"作为约束，这个常数和矛盾一并消失。

## 5. 建议的修订

1. **优先实现共享缓存**（`worker/v2/attn_utils.py` 建 batch 作用域缓存 + `ops/gdn_attn_builder.py` 加命中分支）。预期 17ms → ~5ms，零 KV 代价，不碰 core，不触及 scheduler 契约。**单这一条即超额完成两份文档共同的"p50 下降 ≥50%"门槛。**
2. **分组改动降为可选第二步**，目标改为 `g=8`，规则用"组数最少 subject to padding ≤ 现有 heuristic"，**不做 DSpark gating**。这样同时去掉 padding 预算常数、20%/10% 矛盾、以及头号一致性风险。
3. 两者叠加约 ~3.3ms（GDN ~1.5ms + 非 GDN ~1.8ms），仍是零 KV 代价。

两份文档的测试设计（core 单测、KV 配置构造测试、Ascend metadata 测试）、计时点、NPU 端到端验证流程**基本可原样复用**，只需把"构建单元 15→5"这条门槛换成"GDN 有效完整构建次数 10→1"。容量门槛在共享缓存方案下变成"必须完全一致"，反而更好验。

## 6. 实施与复核提示

- 共享缓存落地时建议加临时断言：同一步内所有 group 的 GDN metadata，除 `non_spec_state_indices_tensor` / `non_spec_conv1d_cache_indices` / spec 两个索引张量外，其余字段应逐位相同。若某字段实际随 group 变化而被误判为共享，断言当场抓住，比事后排查精度问题便宜得多。
- 缓存键需包含真正影响结果的输入（`num_reqs`、`num_tokens`、spec decode 相关 tensor 的身份），避免跨 step 误命中。
- **版本差异提醒**：dsv4 review 引用 `use_dspark()` 位于 `vllm/config/speculative.py:1333`，但本工作区的 vLLM 中该方法在 **1463** 行。若按计划实施，请以实际 runtime 的 vLLM 版本为准复核所有引用行号。

## 7. 需要澄清的一点

本评审的支点是 GDN ≈ 14ms / 总 17ms 这个实测值，来源是单次 profiling。建议补充测量条件（batch size、序列长度分布、是否稳态、eager 还是入图）并附原始 profiling 数据。**若该数字不成立，第 1、2、5 节的结论都需重做。**
